Mathlib/RingTheory/DedekindDomain/Factorization.lean

```diff
@@ -177,3 +223,289 @@ theorem finprod_heightOneSpectrum_factorization_coe (I : Ideal R) (hI : I ≠ 0)
+    let a := Classical.choose (exists_eq_spanSingleton_mul I)
+    let J := Classical.choose (Classical.choose_spec (exists_eq_spanSingleton_mul I))
+    ((Associates.mk v.asIdeal).count (Associates.mk J).factors -
+        (Associates.mk v.asIdeal).count (Associates.mk (Ideal.span {a})).factors : ℤ)
+
+/-- `val_v(I)` does not depend on the choice of `a` and `J` used to represent `I`. -/
+theorem count_well_defined {I : FractionalIdeal (nonZeroDivisors R) K} (hI : I ≠ 0) {a : R}
+    {J : Ideal R} (h_aJ : I = spanSingleton (nonZeroDivisors R) ((algebraMap R K) a)⁻¹ * ↑J) :
+    count K v I = ((Associates.mk v.asIdeal).count (Associates.mk J).factors -
+      (Associates.mk v.asIdeal).count (Associates.mk (Ideal.span {a})).factors : ℤ) := by
+  set a₁ := Classical.choose (exists_eq_spanSingleton_mul I)
+  set J₁ := Classical.choose (Classical.choose_spec (exists_eq_spanSingleton_mul I))
+  have h_a₁J₁ : I = spanSingleton (nonZeroDivisors R) ((algebraMap R K) a₁)⁻¹ * ↑J₁ :=
+    (Classical.choose_spec (Classical.choose_spec (exists_eq_spanSingleton_mul I))).2
+  have h_a₁_ne_zero : a₁ ≠ 0 :=
+    (Classical.choose_spec (Classical.choose_spec (exists_eq_spanSingleton_mul I))).1
+  have h_J₁_ne_zero : J₁ ≠ 0 := ideal_factor_ne_zero hI h_a₁J₁
+  have h_a_ne_zero : Ideal.span {a} ≠ 0 := constant_factor_ne_zero hI h_aJ
+  have h_J_ne_zero : J ≠ 0 := ideal_factor_ne_zero hI h_aJ
+  have h_a₁' : spanSingleton (nonZeroDivisors R) ((algebraMap R K) a₁) ≠ 0 := by
+    rw [Ne.def, spanSingleton_eq_zero_iff, ← (algebraMap R K).map_zero,
+      Injective.eq_iff (IsLocalization.injective K (le_refl (nonZeroDivisors R)))]
+    exact h_a₁_ne_zero
+  have h_a' : spanSingleton (nonZeroDivisors R) ((algebraMap R K) a) ≠ 0 := by
+    rw [Ne.def, spanSingleton_eq_zero_iff, ← (algebraMap R K).map_zero,
+      Injective.eq_iff (IsLocalization.injective K (le_refl (nonZeroDivisors R)))]
+    rw [Ne.def, Ideal.zero_eq_bot, Ideal.span_singleton_eq_bot] at h_a_ne_zero
+    exact h_a_ne_zero
+  have hv : Irreducible (Associates.mk v.asIdeal) := by
+    exact (Associates.irreducible_mk _).mpr v.irreducible
+  rw [h_a₁J₁, ← div_spanSingleton, ← div_spanSingleton, div_eq_div_iff h_a₁' h_a',
+    ← coeIdeal_span_singleton, ← coeIdeal_span_singleton, ← coeIdeal_mul, ← coeIdeal_mul] at h_aJ
+  rw [count, dif_neg hI, sub_eq_sub_iff_add_eq_add, ← ofNat_add, ← ofNat_add, coe_nat_inj',
+    ← Associates.count_mul _ _ hv, ← Associates.count_mul _ _ hv, Associates.mk_mul_mk,
+    Associates.mk_mul_mk, coeIdeal_injective h_aJ]
+  · rw [Ne.def, Associates.mk_eq_zero]; exact h_J_ne_zero
+  · rw [Ne.def, Associates.mk_eq_zero, Ideal.zero_eq_bot, Ideal.span_singleton_eq_bot]
+    exact h_a₁_ne_zero
+  · rw [Ne.def, Associates.mk_eq_zero]; exact h_J₁_ne_zero
+  · rw [Ne.def, Associates.mk_eq_zero]; exact h_a_ne_zero
+
+/-- For nonzero `I, I'`, `val_v(I*I') = val_v(I) + val_v(I')`. -/
+theorem count_mul {I I' : FractionalIdeal (nonZeroDivisors R) K} (hI : I ≠ 0) (hI' : I' ≠ 0) :
```

Comment:
I think this should be generalized to `Finset.prod`.

Reply:
Do you need this? In my applications I always use finprods over all the maximal ideals, so I only needed the finprod version.

Reply:
Say you have a factorization $f(x) = \prod g_i(x)$ of polynomials, then you would get $(f(a)) = \prod (g_i(a))$ where the ideals on the LHS aren't maximal ideals, but you would likely want to show $\text{val}_v((f(a))) = \sum \text{val}_v((g_i(a)))$.   I think these would come up here and there?